### PR TITLE
Remove ExpressionEquivalence

### LINKED ISSUE #21235: Regression: Sql planner fails with contract violation error
The following query fails in `443` 
```
CREATE TABLE memory.default.tbl AS SELECT '' c1, '' column_name_long2, '' new_column_name_long3
;

EXPLAIN
SELECT true
FROM (SELECT c1, column_name_long2 c2, new_column_name_long3 c3 FROM memory.default.tbl)
WHERE (
           c1 LIKE '01'
        OR c1 LIKE '02'
        OR c1 LIKE '03'
        OR c1 LIKE '04'
        OR c1 LIKE '05'
        OR c1 LIKE '06'
        OR c1 LIKE '07'
        OR c1 LIKE '08'
        OR c1 LIKE '09'
        OR c1 LIKE '10'
        OR c1 LIKE '11'
        OR c1 LIKE '12'
        OR c1 LIKE '13'
        OR c1 LIKE '14'
        OR c1 LIKE '15'
        OR c1 LIKE '16'
        OR c1 LIKE '17'
        OR c1 LIKE '18'
        OR c1 LIKE '19'
        OR c1 LIKE '20'
        OR c1 LIKE '21'
        OR c1 LIKE '22'
        OR c1 LIKE '23'
        OR c1 LIKE '24'
        OR c2 LIKE '25'
        OR c2 LIKE '26'
        OR c2 LIKE '27'
        OR c2 LIKE '28'
        OR c3 LIKE '29'
        OR c3 LIKE '30'
        OR c3 LIKE '31'
        OR c2 LIKE '32'
    )
;
```

Error stack trace is:
```
java.lang.IllegalArgumentException: Comparison method violates its general contract!
	at java.base/java.util.TimSort.mergeHi(TimSort.java:903)
	at java.base/java.util.TimSort.mergeAt(TimSort.java:520)
	at java.base/java.util.TimSort.mergeForceCollapse(TimSort.java:461)
	at java.base/java.util.TimSort.sort(TimSort.java:254)
	at java.base/java.util.Arrays.sort(Arrays.java:1234)
	at com.google.common.collect.Ordering.sortedCopy(Ordering.java:871)
	at io.trino.sql.planner.optimizations.ExpressionEquivalence$CanonicalizationVisitor.visitSpecialForm(ExpressionEquivalence.java:146)
	at io.trino.sql.planner.optimizations.ExpressionEquivalence$CanonicalizationVisitor.visitSpecialForm(ExpressionEquivalence.java:100)
	at io.trino.sql.relational.SpecialForm.accept(SpecialForm.java:144)
	at io.trino.sql.planner.optimizations.ExpressionEquivalence.areExpressionsEquivalent(ExpressionEquivalence.java:82)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.areExpressionsEquivalent(PredicatePushDown.java:1188)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitFilter(PredicatePushDown.java:383)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitFilter(PredicatePushDown.java:145)
	at io.trino.sql.planner.plan.FilterNode.accept(FilterNode.java:74)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:423)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitProject(PredicatePushDown.java:278)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitProject(PredicatePushDown.java:145)
	at io.trino.sql.planner.plan.ProjectNode.accept(ProjectNode.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.rewrite(SimplePlanRewriter.java:81)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.lambda$defaultRewrite$0(SimplePlanRewriter.java:72)
	at com.google.common.collect.ImmutableList.forEach(ImmutableList.java:423)
	at io.trino.sql.planner.plan.SimplePlanRewriter$RewriteContext.defaultRewrite(SimplePlanRewriter.java:72)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitPlan(PredicatePushDown.java:182)
	at io.trino.sql.planner.optimizations.PredicatePushDown$Rewriter.visitPlan(PredicatePushDown.java:145)
	at io.trino.sql.planner.plan.PlanVisitor.visitOutput(PlanVisitor.java:49)
	at io.trino.sql.planner.plan.OutputNode.accept(OutputNode.java:82)
	at io.trino.sql.planner.plan.SimplePlanRewriter.rewriteWith(SimplePlanRewriter.java:31)
	at io.trino.sql.planner.optimizations.PredicatePushDown.optimize(PredicatePushDown.java:139)
	at io.trino.sql.planner.optimizations.StatsRecordingPlanOptimizer.optimize(StatsRecordingPlanOptimizer.java:41)
	at io.trino.sql.planner.LogicalPlanner.runOptimizer(LogicalPlanner.java:301)
	at io.trino.sql.planner.LogicalPlanner.plan(LogicalPlanner.java:264)
	at io.trino.sql.analyzer.QueryExplainer.getLogicalPlan(QueryExplainer.java:176)
	at io.trino.sql.analyzer.QueryExplainer.getDistributedPlan(QueryExplainer.java:187)
	at io.trino.sql.analyzer.QueryExplainer.getPlan(QueryExplainer.java:106)
	at io.trino.sql.rewrite.ExplainRewrite$Visitor.getQueryPlan(ExplainRewrite.java:151)
	at io.trino.sql.rewrite.ExplainRewrite$Visitor.visitExplain(ExplainRewrite.java:129)
	at io.trino.sql.rewrite.ExplainRewrite$Visitor.visitExplain(ExplainRewrite.java:74)
	at io.trino.sql.tree.Explain.accept(Explain.java:61)
	at io.trino.sql.tree.AstVisitor.process(AstVisitor.java:27)
	at io.trino.sql.rewrite.ExplainRewrite.rewrite(ExplainRewrite.java:71)
	at io.trino.sql.rewrite.StatementRewrite.rewrite(StatementRewrite.java:54)
	at io.trino.sql.analyzer.Analyzer.analyze(Analyzer.java:92)
	at io.trino.sql.analyzer.Analyzer.analyze(Analyzer.java:86)
	at io.trino.execution.SqlQueryExecution.analyze(SqlQueryExecution.java:281)
	at io.trino.execution.SqlQueryExecution.<init>(SqlQueryExecution.java:214)
	at io.trino.execution.SqlQueryExecution$SqlQueryExecutionFactory.createQueryExecution(SqlQueryExecution.java:878)
	at io.trino.dispatcher.LocalDispatchQueryFactory.lambda$createDispatchQuery$0(LocalDispatchQueryFactory.java:153)
	at io.trino.$gen.Trino_443____20240325_125212_2.call(Unknown Source)
	at com.google.common.util.concurrent.TrustedListenableFutureTask$TrustedFutureInterruptibleTask.runInterruptibly(TrustedListenableFutureTask.java:131)
	at com.google.common.util.concurrent.InterruptibleTask.run(InterruptibleTask.java:76)
	at com.google.common.util.concurrent.TrustedListenableFutureTask.run(TrustedListenableFutureTask.java:82)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1583)
```

Same query succeed in `442`

### PR DESCRIPTION
**Pull request by martint**
This abstraction doesn't appear to be needed anymore.

Fixes #21235 

## Release notes

(x) Release notes are required, with the following suggested text:

```markdown
# Section
* TBD. ({issue}`21235`)
```

### CODE REVIEW

**Comment by findepi** [comment]
> Remove TestExpressionEquivalence

Did you mean "TestEE" or "EE"?

**Comment by martint** [comment]
Good catch. Yes, ExpressionEquivalence and associated tests.

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java [hunk 1]
```diff
@@ -764,7 +762,7 @@ public PlanNode visitSpatialJoin(SpatialJoinNode node, RewriteContext<Expression
             PlanNode output = node;
             if (leftSource != node.getLeft() ||
                     rightSource != node.getRight() ||
-                    !areExpressionsEquivalent(newJoinPredicate, joinPredicate)) {
+                    !newJoinPredicate.equals(joinPredicate)) {
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java:765]: so the assumption is that newJoinPredicate cannot be similar to (equivalent), but different from, joinPredicate?

what does guarantee this?

**Comment by findepi** [comment]
LGTM, but i don't think I know why/when it became obsolete.

@sopel39 do you want to take a look?

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java (hunk 1 above)
Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java:765]: This is just a stronger requirement to detect whether the predicate was rewritten by the optimization. Before, it just checked whether they were equivalent based on some semantic-preserving transformations. Now it's based on whether they are exactly the same.

**Comment by martint** [comment]
> LGTM, but i don't think I know why/when it became obsolete.

The optimizer performs many of the canonicalizations that this class was doing. I wasn't able to find any queries that produce different plans with or without this code.

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java (hunk 1 above)
Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java:765]: The question is, was `ExpressionEquivalence` more forgiving than `equals`? Maybe we keep `ExpressionEquivalence` abstraction but with `equals` inside?

Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java:765]: Yes, it was more lenient than equality, however, I don't think that matters in practice.

The only canonicalizations this class supports are:

* `AND(AND(x, y), z)` => `AND(x, y, z)`. Same for `OR`.
* `AND(x, x, x, a)` => `AND(x, a)`. Same for `OR`. I believe there's a bug here, though, as it ignores non-deterministic operations that should be evaluated multiple times.
* `a = b` => `b = a`
* `a is distinct from b` => `b is distinct from a`

All of these are already handled by the expression optimizer and canonicalizer.

Comment (martint) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java:765]: > I believe there's a bug here, though, as it ignores non-deterministic operations that should be evaluated multiple times.

This demonstrates the issue:

```java
ResolvedFunction random = new TestingFunctionResolution().resolveFunction("random", fromTypes());
Expression expression = new IsNull(new Call(random, List.of()));
assertNotEquivalent(
        new Logical(AND, List.of(expression, expression)),
        expression);
```

**Review by dain — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java (hunk 1 above)
Comment (findepi) [reply core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java:765]: > All of these are already handled by the expression optimizer and canonicalizer.

the `newJoinPredicate` comes thru the optimizer (`PredicatePushDown.Rewriter#simplifyExpression`), does it also do canonicalization?

**Pull request merged**